packages/muelu/src/Smoothers/MueLu_BelosSmoother_decl.hpp

```diff
@@ -0,0 +1,249 @@
+#undef MUELU_BELOSSMOOTHER_SHORT
+#include "MueLu_UseShortNames.hpp"
+
+  public:
+
+    //! @name Constructors / destructors
+    //@{
+    //TODO: update doc for Belos.
+    /*! @brief Constructor
+
+    The options passed into BelosSmoother are those given in the Belos user's manual.
+
+    @param type smoother type
+    @param list options for the particular smoother (e.g., fill factor or damping parameter)
+
+    */
+
+#ifndef _MSC_VER
+    // Avoid error C3772: invalid friend template declaration
+    template<class Scalar2, class LocalOrdinal2, class GlobalOrdinal2, class Node2>
+    friend class BelosSmoother;
+#endif
+
+    BelosSmoother(const std::string type, const Teuchos::ParameterList& paramList = Teuchos::ParameterList());
+
+    //! Destructor
+    virtual ~BelosSmoother() { }
+
+    //@}
+
+    void SetParameterList(const Teuchos::ParameterList& paramList);
+
+    //! Input
+    //@{
+
+    void DeclareInput(Level &currentLevel) const;
+
+    //@}
+
+    //! @name Computational methods.
+    //@{
+
+    /*! @brief Set up the smoother.
+
+    This creates the underlying Belos smoother object, copies any parameter list options
+    supplied to the constructor to the Belos object, and computes the preconditioner.
+    */
+    void Setup(Level &currentLevel);
+
+    /*! @brief Apply the preconditioner.
+
+    Solves the linear system <tt>AX=B</tt> using the constructed smoother.
+
+    @param X initial guess
+    @param B right-hand side
+    @param InitialGuessIsZero (optional) If false, some work can be avoided. Whether this actually saves any work depends on the underlying Belos implementation.
+    */
+    void Apply(MultiVector& X, const MultiVector& B, bool InitialGuessIsZero = false) const;
+
+    //@}
+
+    //! @name Utilities
+    //@{
+
+    RCP<SmootherPrototype> Copy() const;
+
+    //@}
+
+    //! @name Overridden from Teuchos::Describable
+    //@{
+
+    //! Return a simple one-line description of this object.
+    std::string description() const;
+
+    //! Print the object with some verbosity level to an FancyOStream object.
+    //using MueLu::Describable::describe; // overloading, not hiding
+    //void describe(Teuchos::FancyOStream &out, const VerbLevel verbLevel = Default) const
+    void print(Teuchos::FancyOStream &out, const VerbLevel verbLevel = Default) const;
+
+    //@}
+
+    //! For diagnostic purposes
+    // RCP<Ifpack2::Preconditioner<Scalar,LocalOrdinal,GlobalOrdinal,Node> > getPreconditioner(){return prec_;}
+
+    //! Get a rough estimate of cost per iteration
+    size_t getNodeSmootherComplexity() const;
+
+  private:
+    void SetupBelos           (Level& currentLevel);
+
+  private:
+
+    std::string                         type_;
+    // typedef Tpetra::Operator<Scalar>             OP;
+    // typedef Tpetra::MultiVector<Scalar>          MV;
+    typedef MultiVector          MV;
+    typedef Belos::OperatorT<MV> OP;
+
+    RCP<Belos::LinearProblem<Scalar, MV, OP> > belosProblem_;
```

Comment:
It would be better to dispatch to the Epetra or Tpetra specialization of Belos, instead of using Xpetra or this `Belos::OperatorT` thing.  Right now, you're forcing Trilinos to rebuild all of Belos.

Reply:
That should allow me to get all the solvers that are registered for Tpetra/Epetra respectively, right? Cause I noticed that some of the solvers were supposedly supported (via `supportedSolverNames`), but not via `isSupported`, which is somewhat odd behavior.

Reply:
@cgcgcg wrote:
> That should allow me to get all the solvers that are registered for Tpetra/Epetra respectively, right?  Cause I noticed that some of the solvers were supposedly supported (via `supportedSolverNames`), but not via `isSupported`, which is somewhat odd behavior.

Would you happen to have an example?  Both of these functions appear to check the custom factories, which would let you call things like the Tpetra-only communication-avoiding and pipelined solvers.